backends/cuda/ceed-cuda-vector.c

```diff
@@ -257,89 +390,69 @@ static int CeedVectorSetValue_Cuda(CeedVector vec, CeedScalar val) {
 //   this will perform a copy (possibly cached).
 //------------------------------------------------------------------------------
 static int CeedVectorGetArrayRead_Cuda(const CeedVector vec,
-                                       const CeedMemType mtype,
-                                       const CeedScalar **array) {
+                                       const CeedMemType mtype, const CeedScalar **array) {
   int ierr;
   Ceed ceed;
   ierr = CeedVectorGetCeed(vec, &ceed); CeedChkBackend(ierr);
-  CeedVector_Cuda *data;
-  ierr = CeedVectorGetData(vec, &data); CeedChkBackend(ierr);
+  CeedVector_Cuda *impl;
+  ierr = CeedVectorGetData(vec, &impl); CeedChkBackend(ierr);
 
-  // Sync array to requested memtype and update pointer
+  bool has_valid_array = true, is_array_of_type = true;
+  ierr = CeedVectorHasValidArray(vec, &has_valid_array); CeedChkBackend(ierr);
+  ierr = CeedVectorIsArrayOfType_Cuda(vec, mtype, &is_array_of_type);
+  CeedChkBackend(ierr);
+  if (!has_valid_array && is_array_of_type) {
+    // LCOV_EXCL_START
```

Comment:
I don't understand why we want `is_array_of_type` here.  It seems that `is_array_of_type` will be true if either of the `borrowed` or `owned` arrays exists for the requested memtype, correct?  Why do we want to check specifically that the vector does *not* have a valid array but *does* have some invalid/stale data for the memtype before throwing an error?  Isn't no valid data alone enough to throw an error for `GetArrayRead`?

Reply:
We use the pattern of calling GetArray before setting any allocations all over the place.

```
CeedVectorCreate(ceed, 10, &vec);
CeedVectorGetArray(vec, &array);
for (i=0; i<10; i++) {
  // code initializing values
}
```

I want to support this while preventing access to stale data.

Reply:
Let me see if I understand now: the goal is basically to make sure that the vector either a) has valid data, or b) doesn't have *any* data (for the memtype) at all?  Otherwise, if it has data but the data is stale, it is an error?  

My understanding from our discussion of the future multiprecision vectors was that once a vector is given access to a memory location through `SetArray` + `USE_POINTER`, it should continue to assume it can use that memory location until told otherwise (through `TakeArray` or a new `SetArray` call).  For `GetArrayRead`, it makes sense that we wouldn't want access to stale data, since we presumably care about the current values in the vector (otherwise, why use `GetArrayRead`?).  But for `GetArray`, couldn't we want access to the stale data *location* in order to write new values (and then make the data valid again)?  

Actually, this leads me to another question: if we presume we can use the borrowed location until told otherwise, and we also end up allocating memory later, which one (owned or borrowed) do we prefer later, if neither is currently valid?  ...Could such a scenario even ever happen?

Reply:
> the goal is basically to make sure that the vector either a) has valid data, or b) doesn't have any data (for the memtype) at all? Otherwise, if it has data but the data is stale, it is an error?

Exactly.

>  But for `GetArray`, couldn't we want access to the stale data location in order to write new values (and then make the data valid again)?

For `GetArray` I am treating it as read/write access. I could see a user wanting to read some values and write some values, which would lead to an array with a mix of valid and invalid data.

We would only fall into this use case if the user Sets the array and then Takes the array back without ensuring synchronization. I would expect a user to call `SyncArray' before taking the pointer back if they want to make sure the data is valid in a particular memory location that they are not taking or to call `SetValue` to reset the validity of the data at the location.

> if we presume we can use the borrowed location until told otherwise, and we also end up allocating memory later, which one (owned or borrowed) do we prefer later, if neither is currently valid? ...Could such a scenario even ever happen?

I think we should try to prevent this from ever happening. Setting a borrowed array should free a corresponding allocation and vise versa.

Reply:
>  I would expect a user to call `SyncArray` before taking the pointer back if they want to make sure the data is valid in a particular memory location that they are not taking or to call `SetValue` 

Wouldn't this be a departure from how it currently works, though?  Didn't we do the checking for them and sync if necessary when requesting memory access on host/device, no explicit `Sync` required?

> I could see a user wanting to read some values and write some values, which would lead to an array with a mix of valid and invalid data.

But in this setup, `GetArray` and `GetArrayRead` work the same way, so `GetArray` would also error if you were trying to get access to stale data -- but it shouldn't, right?  `GetArray` should let you access stale data for write purposes?

Actually, this is something I was already wondering about in terms of multiprecision -- should we have a special case for write-only access, where we know we don't care about syncing the current values even if they are stale?  And a version that will do the sync (for host/device now, for precision later) if required?

Reply:
I have no objection to adding a `GetArrayInvalid` where the user requests access to a dirty array, but it isn't something that we currently need internally. With the exception of the one small bug I found in the `/cpu/self/opt` backends, internally and in the examples we have only been calling `GetArray` if we have valid data or no arrays yet allocated.

Reply:
PETSc has `VecGetArrayWrite()`, which makes it clear at the call site that old values are to be considered invalid. In debug mode, these can be initialized with NaN to prevent accidental use.

Reply:
That name makes sense to me. I added this function to the interface.

We could adopt the convention that `GetArray` and `GetArrayRead` require valid data while `GetArrayWrite` is the only function that allows access to uninitialized or invalid data. That would simplify the error checking, but would require a couple of small changes in the backends. Any votes yea on nay?

Reply:
> We could adopt the convention that GetArray and GetArrayRead require valid data while GetArrayWrite is the only function that allows access to uninitialized or invalid data. 

I like it, yea vote from me 👍

Reply:
I still do think this may cause minor changes somewhere with usage.  Previously, if you called `GetArray` or `GetArrayRead` with device memtype, and the host data was set, the vector would sync before giving you the device pointer.  Now, it would cause an error unless you explicitly call `Sync` first, is that correct?

Reply:
Not quite. I think that automatic synchronization should be expected, unless there exists no valid data to synchronize. If there is no valid data, then in this branch read or read/write access will fail.

This should be pretty uncommon. The  only use case I can think of would be if the user calls `TakeArray` on the only valid memory location and then wants to get read access to the array.